File: eclipse-initial/src/native/org.eclipse.ice.reactor.native/LWRComposite.cpp
/*******************************************************************************
 * Copyright (c) 2012, 2014 UT-Battelle, LLC.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Initial API and implementation and/or initial documentation - Jay Jay Billings,
 *   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
 *   Claire Saunders, Matthew Wang, Anna Wojtowicz
 *******************************************************************************/

#include "LWRComposite.h"
#include <IHdfReadable.h>
#include <IHdfWriteable.h>
#include <ICEObject/Identifiable.h>
#include <stdexcept>
#include <map>
#include <vector>
#include <string>

using namespace ICE_Reactor;

/**
 * The Copy constructor.
 *
 * @param The object to be copied.
 */
LWRComposite::LWRComposite(LWRComposite & arg) : LWRComponent(arg) {

    //Copy contents

    //Deep copy tree
    this->lWRComponents.clear();

    std::map<std::string, std::shared_ptr <LWRComponent> >::iterator mapIter;

    //Iterate the map and add to list
    for(mapIter = arg.lWRComponents.begin(); mapIter != arg.lWRComponents.end(); ++mapIter) {
        std::shared_ptr<LWRComponent> component ( new LWRComponent(*mapIter->second.get()));
        this->lWRComponents.insert( std::pair<std::string, std::shared_ptr <LWRComponent> >(mapIter->first, component) );
    }
}

/**
 * The Destructor
 */
LWRComposite::~LWRComposite() {
    //TODO Auto-generated method stub
}

/**
 * The Constructor.
 */
LWRComposite::LWRComposite() : LWRComponent() {

    //Setup default values
    this->name = "Composite 1";
    this->id = 1;
    this->description = "Composite 1's Description";
    this->HDF5LWRTag = LWRCOMPOSITE;

}

/**
 * Returns the LWRComponent corresponding to the provided name or null if the name is not found.
 *
 * @param The name
 *
 * @return the component
 */
std::shared_ptr<ICE_DS::Component> LWRComposite::getComponent(const std::string name) {

    //Local declarations
    std::shared_ptr<ICE_DS::Component> nullComponent;
    std::map<std::string, std::shared_ptr <LWRComponent> >::iterator mapIter;

    //Iterate the map and add to list
    for(mapIter = this->lWRComponents.begin(); mapIter != this->lWRComponents.end(); ++mapIter) {

        if(mapIter->second.get()->getName().compare(name) == 0) {
            return mapIter->second;
        }

    }

    //Id did not exist
    return nullComponent;
}

/**
 * Returns a list of std::strings containing the names of all LWRComponents contained in this LWRComposite.
 *
 * @return the list of strings
 */
std::vector<std::string> LWRComposite::getComponentNames() {

    // begin-user-code

    //Local declarations
    std::vector<std::string> list;
    std::map<std::string, std::shared_ptr <LWRComponent> >::iterator mapIter;

    //Iterate the map and add to list
    for(mapIter = this->lWRComponents.begin(); mapIter != this->lWRComponents.end(); ++mapIter) {
        list.push_back(mapIter->first);
    }

    //Return the list
    return list;

    // end-user-code
}

/**
 * Removes a LWRComponent with the provided name from this LWRComposite.
 *
 * @param the name
 */
void LWRComposite::removeComponent(const std::string name) {
    // begin-user-code

    //Local declarations
    std::vector<std::string> list;
    std::map<std::string, std::shared_ptr <LWRComponent> >::iterator mapIter;

    //Iterate the map and remove item if equal
    for(mapIter = this->lWRComponents.begin(); mapIter != this->lWRComponents.end(); ++mapIter) {

        //If the name exists, remove it
        if(name.compare(mapIter->first) == 0) {
            this->lWRComponents.erase(mapIter->first);
            break;
        }
    }

    //Return
    return;
}

/**
 * Overrides the equals operation to check the attributes on this object with another object of the same type.  Returns true if the objects are equal.  False otherwise.
 *
 * @param object to compare to
 *
 * @return true if equal, false otherwise
 */
bool LWRComposite::operator ==(const LWRComposite& otherObject) const {

    // begin-user-code

    //Check map values
    std::map<std::string, std::shared_ptr <LWRComponent> >::const_iterator mapIter;

    //Check super equality and the map sizes
    if(LWRComponent::operator ==(otherObject) == false || this->lWRComponents.size() != otherObject.lWRComponents.size() ) return false;

    //Iterate the map and check values
    for(mapIter = this->lWRComponents.begin(); mapIter != this->lWRComponents.end(); ++mapIter) {

        //IF the values do not compare correctly, return false
        try {
            if(this->lWRComponents.at(mapIter->first).get()->operator ==(*otherObject.lWRComponents.at(mapIter->first).get()) == false) {
                return false;
            }
        } catch (const std::out_of_range& oor) {
            return false;
        }

    }


    //everything true, return true
    return true;

    // end-user-code

}

/**
 * Deep copies and returns a newly instantiated object.
 *
 * @return The newly instantiated object.
 */
std::shared_ptr<ICE_DS::Identifiable> LWRComposite::clone() {
    // begin-user-code

    //Local Declarations
    std::shared_ptr<LWRComponent> component (new LWRComposite(*this));

    //Return the component
    return component;

    // end-user-code
}

/**
 * Returns the list of writeable childen.
 *
 * @return List of writeable chidlren
 */
std::vector< std::shared_ptr<ICE_IO::IHdfWriteable> > LWRComposite::getWriteableChildren() {


    std::vector< std::shared_ptr<ICE_IO::IHdfWriteable> > writeables;
    std::map<std::string, std::shared_ptr <LWRComponent> >::const_iterator mapIter;

    //Loop over the map and get the values and then add it to a vector
    for(mapIter = this->lWRComponents.begin(); mapIter != this->lWRComponents.end(); ++mapIter) {

        writeables.push_back(mapIter->second);
    }

    return writeables;


}

/**
 * This operation reads IHdfReadable child objects. If this IHdfReadable has no IHdfReadable child objects, false is returned.
 *
 * @param the readable child
 *
 * @return true if successful, false otherwise
 */
bool LWRComposite::readChild(std::shared_ptr<ICE_IO::IHdfReadable> iHdfReadable) {

    // begin-user-code

    //If the child is null, return false
    if(iHdfReadable.get() == NULL) {
        return false;
    }

    //Cast the child into a LWRComponent
    std::shared_ptr<LWRComponent> childComponent = std::dynamic_pointer_cast<LWRComponent> (iHdfReadable);

    //Remove any child with the new child's name from this LWRComposite
    LWRComposite::removeComponent(childComponent.get()->getName());

    //Add the new child LWRComponent
    LWRComposite::addComponent(childComponent);

    //Return true for success
    return true;

    // end-user-code
}

/**
 * This operation adds a child Component to a class that realizes the Composite interface. This operation should notify listeners that components have been added.
 *
 * @param the child to add
 */
void LWRComposite::addComponent(std::shared_ptr<ICE_DS::Component> child) {

    if(child.get() == NULL) {
        return;
    }

    //If found, return null
    std::map<std::string, std::shared_ptr <LWRComponent> >::iterator mapIter;

    //Iterate the map and remove item if equal
    for(mapIter = this->lWRComponents.begin(); mapIter != this->lWRComponents.end(); ++mapIter) {

        //If the name exists, return
        if(child.get()->getName().compare(mapIter->first) == 0) {
            return;
        }
    }

    //Downcast to a shared pointer of LWRComponent type

    //Cast up
    std::shared_ptr<LWRComponent> castedComponent = std::dynamic_pointer_cast <LWRComponent> (child);

    //Name doesnt exist, add it!
    this->lWRComponents.insert( std::pair<std::string, std::shared_ptr <LWRComponent> >(child.get()->getName(), castedComponent) );

}

/**
 * This operation removes a child Component to a class that realizes the Composite interface.
 *
 * @param the child id
 */

void LWRComposite::removeComponent(int childId) {
    // begin-user-code

    //Local declarations
    std::vector<std::string> list;
    std::map<std::string, std::shared_ptr <LWRComponent> >::iterator mapIter;

    //Iterate the map and remove item if equal
    for(mapIter = this->lWRComponents.begin(); mapIter != this->lWRComponents.end(); ++mapIter) {

        //If the name exists, remove it
        if(mapIter->second.get()->getId() == childId) {
            this->lWRComponents.erase(mapIter->first);
            break;
        }
    }

    //Return
    return;
}

/**
 * This operation retrieves a child Component to a class that realizes the Composite interface.
 *
 * @param the child id
 *
 * @return the component
 */
std::shared_ptr<ICE_DS::Component> LWRComposite::getComponent(int childId) {

    //begin-user-code

    //Local declarations
    std::shared_ptr<ICE_DS::Component> nullComponent;

    std::map<std::string, std::shared_ptr <LWRComponent> >::iterator mapIter;

    //Iterate the map and add to list
    for(mapIter = this->lWRComponents.begin(); mapIter != this->lWRComponents.end(); ++mapIter) {
        if(mapIter->second.get()->getId() == childId) {
            //remove, break
            return this->lWRComponents.at(mapIter->first);
            break;
        }

    }

    //Id did not exist
    return nullComponent;

}

/**
 * This operations retrieves the number of child Components stored in an instance of a class that realizes the Composite interface.
 *
 * @return the number of components
 */
int LWRComposite::getNumberOfComponents() {
    return this->lWRComponents.size();
}


/**
 * This operation returns all of the Components stored in the Composite.
 *
 * @return the list of components
 */
std::vector< std::shared_ptr<ICE_DS::Component> > LWRComposite::getComponents() {

    //begin-user-code

    //Local declarations
    std::vector< std::shared_ptr<ICE_DS::Component> > components;

    std::map<std::string, std::shared_ptr <LWRComponent> >::iterator mapIter;

    //Iterate the map and add to list
    for(mapIter = this->lWRComponents.begin(); mapIter != this->lWRComponents.end(); ++mapIter) {

        //Push to the list
        components.push_back(mapIter->second);

    }

    //return list
    return components;

}

File: src/org.eclipse.ice.reactor/src/org/eclipse/ice/reactor/LWRComposite.java
/*******************************************************************************
 * Copyright (c) 2012, 2014 UT-Battelle, LLC.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Initial API and implementation and/or initial documentation - Jay Jay Billings,
 *   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
 *   Claire Saunders, Matthew Wang, Anna Wojtowicz
 *******************************************************************************/
package org.eclipse.ice.reactor;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.Iterator;

import javax.xml.bind.annotation.XmlTransient;

import org.eclipse.ice.datastructures.ICEObject.Component;
import org.eclipse.ice.datastructures.ICEObject.Composite;
import org.eclipse.ice.io.hdf.IHdfReadable;
import org.eclipse.ice.io.hdf.IHdfWriteable;

/**
 * <p>
 * The LWRComposite class represents all reactor components that can store and
 * manage LWRComponents. This class implements the ICE Composite interface. This
 * class was designed as a "branch" within the Reactor package to hold
 * references to other LWRComponents. Although this class implements the
 * Composite interface, classes that extend LWRComposite should consider if they
 * will need to override the Composite Interface's operations to provide
 * specific utility as needed.
 * </p>
 * 
 * @author Scott Forest Hull II
 */
public class LWRComposite extends LWRComponent implements Composite {
	/**
	 * <p>
	 * A Hashtable keyed on LWRComponent name storing unique LWRComponents.
	 * </p>
	 * 
	 */
	@XmlTransient
	protected Hashtable<String, LWRComponent> lWRComponents;

	/**
	 * <p>
	 * The Constructor.
	 * </p>
	 * 
	 */
	public LWRComposite() {

		// Setup the composite
		this.name = "Composite 1";
		this.id = 1;
		this.description = "Composite 1's Description";

		// Setup hashTable
		this.lWRComponents = new Hashtable<String, LWRComponent>();

		// Setup the LWRComponentType to correct type
		this.HDF5LWRTag = HDF5LWRTagType.LWRCOMPOSITE;

	}

	/**
	 * <p>
	 * Returns the LWRComponent corresponding to the provided name or null if
	 * the name is not found.
	 * </p>
	 * 
	 * @param name
	 *            <p>
	 *            The provided LWRComponent's name.
	 *            </p>
	 * @return <p>
	 *         The LWRComponent corresponding to the provided name or null if
	 *         the name is not found.
	 *         </p>
	 */
	public LWRComponent getComponent(String name) {

		// if the name is not null, return the name
		if (name != null) {
			return this.lWRComponents.get(name);
		}

		// Name was null, return
		return null;
	}

	/**
	 * <p>
	 * Returns an ArrayList of Strings containing the names of all LWRComponents
	 * contained in this LWRComposite.
	 * </p>
	 * 
	 * @return <p>
	 *         An ArrayList of Strings containing the names of all LWRComponents
	 *         contained in this LWRComposite
	 *         </p>
	 */
	public ArrayList<String> getComponentNames() {
		// Local Declarations
		ArrayList<String> keys = new ArrayList<String>();
		Iterator<String> iter;
		String keyName;

		// Setup iterator
		iter = this.lWRComponents.keySet().iterator();

		// Iterate over the list and get the component
		// Only grab the first in the list
		while (iter.hasNext()) {

			// Iterate the next piece
			keyName = iter.next();
			// Add each string key to the list
			keys.add(keyName);
		}

		// Return the names
		return keys;
	}

	/**
	 * <p>
	 * Removes a LWRComponent with the provided name from this LWRComposite.
	 * </p>
	 * 
	 * @param name
	 *            <p>
	 *            The name of the LWRComponent to remove.
	 *            </p>
	 */
	public void removeComponent(String name) {
		// Local Declarations

		// IF the name is not null and it is contained in the list (trimmed),
		// then remove it
		if (name != null && this.lWRComponents.containsKey(name)
				&& !name.trim().isEmpty()) {
			this.lWRComponents.remove(name);
		}
	}

	/**
	 * <p>
	 * Overrides the equals operation to check the attributes on this object
	 * with another object of the same type. Returns true if the objects are
	 * equal. False otherwise.
	 * </p>
	 * 
	 * @param otherObject
	 *            <p>
	 *            The object to be compared.
	 *            </p>
	 * @return <p>
	 *         True if otherObject is equal. False otherwise.
	 *         </p>
	 */
	@Override
	public boolean equals(Object otherObject) {

		// Local declarations
		LWRComposite composite;
		boolean retVal = false;

		// If this object is an instance of the LWRComponent, cast it.
		// Make sure it is also not null
		if (otherObject != null && otherObject instanceof LWRComposite) {
			composite = (LWRComposite) otherObject;

			// If they are on the same heap, return true
			if (this == otherObject) {
				return true;
			}
			// Check values
			retVal = (super.equals(otherObject) && this.lWRComponents
					.equals(composite.lWRComponents));

		}

		// Return retVal
		return retVal;

	}

	/**
	 * <p>
	 * Returns the hashCode of the object.
	 * </p>
	 * 
	 * @return <p>
	 *         The hash of the object.
	 *         </p>
	 */
	@Override
	public int hashCode() {

		// Get the super's hashCode
		int hash = super.hashCode();

		// Hash the hashTable
		hash += 31 * this.lWRComponents.hashCode();

		// Return the hashCode
		return hash;

	}

	/**
	 * <p>
	 * Deep copies the contents of the object.
	 * </p>
	 * 
	 * @param otherObject
	 *            <p>
	 *            The object to be copied.
	 *            </p>
	 */
	public void copy(LWRComposite otherObject) {

		// Local declarations
		Iterator<String> iter;

		// If the object is null, return
		if (otherObject == null) {
			return;
		}
		// Copy super contents
		super.copy(otherObject);

		// Copy the HashTable's contents
		this.lWRComponents.clear();
		iter = otherObject.lWRComponents.keySet().iterator();

		while (iter.hasNext()) {
			String key = iter.next();
			this.lWRComponents.put(key,
					(LWRComponent) otherObject.lWRComponents.get(key));

		}

	}

	/**
	 * <p>
	 * Deep copies and returns a newly instantiated object.
	 * </p>
	 * 
	 * @return <p>
	 *         The newly instantiated copied object.
	 *         </p>
	 */
	@Override
	public Object clone() {

		// Local Declarations
		LWRComposite composite = new LWRComposite();

		// Copy contents
		composite.copy(this);

		// Return the newly instantiated object
		return composite;

	}

	/**
	 * Gets the children of this instance that can be written to the HDF format.
	 * 
	 * @return A list containing the writable children.
	 */
	@Override
	public ArrayList<IHdfWriteable> getWriteableChildren() {

		// Get the children in super
		ArrayList<IHdfWriteable> children = super.getWriteableChildren();

		// If super had no children
		if (children == null) {

			// Initialize to new array list
			children = new ArrayList<IHdfWriteable>();
		}

		// Add all of these children
		children.addAll(lWRComponents.values());

		return children;
	}

	/**
	 * <p>
	 * This operation returns an ArrayList of IHdfReadable child objects. If
	 * this IHdfReadable has no IHdfReadable child objects, then null is
	 * returned.
	 * </p>
	 * 
	 * @param iHdfReadable
	 *            The HDF child of this instance.
	 * @return True if the child could be read and added, false otherwise.
	 */
	@Override
	public boolean readChild(IHdfReadable iHdfReadable) {

		// If the child is null or not an instance of LWRComponent, then return
		// false.
		if (iHdfReadable == null || !(iHdfReadable instanceof LWRComponent)) {
			return false;
		}

		// Cast the child into a LWRComponent
		LWRComponent childComponent = (LWRComponent) iHdfReadable;

		// Remove any child with the new child's name from this LWRComposite
		this.removeComponent(childComponent.getName());

		// Add the new child LWRComponent
		this.addComponent(childComponent);

		// Return true for success
		return true;
	}

	/**
	 * (non-Javadoc)
	 * 
	 * @see Composite#addComponent(Component child)
	 */
	@Override
	public void addComponent(Component child) {

		// If the child is null or if the name is already in the table, return
		if (child == null || this.lWRComponents.containsKey(child.getName())
				|| !(child instanceof LWRComponent)) {
			return;
		}

		// Otherwise, add component
		this.lWRComponents.put(child.getName(), (LWRComponent) child);

	}

	/**
	 * (non-Javadoc)
	 * 
	 * @see Composite#removeComponent(int childId)
	 */
	@Override
	public void removeComponent(int childId) {
		// Local declarations
		Component component = null;
		Iterator<LWRComponent> iter;

		// Setup iterator
		iter = this.lWRComponents.values().iterator();

		// Iterate over the list and get the component
		// Only grab the first in the list
		while (iter.hasNext()) {

			// Iterate the next piece
			component = iter.next();

			// If the component's id matches the required child's id, remove the
			// component
			if (component.getId() == childId) {
				this.removeComponent(component.getName());
			}
		}

	}

	/**
	 * (non-Javadoc)
	 * 
	 * @see Composite#getComponent(int childId)
	 */
	@Override
	public Component getComponent(int childId) {
		// Local declarations
		Component component = null;
		Iterator<LWRComponent> iter;

		// Setup iterator
		iter = this.lWRComponents.values().iterator();

		// Iterate over the list and get the component
		// Only grab the first in the list
		while (iter.hasNext()) {

			// Iterate the next piece
			component = iter.next();

			// If the component's id matches the required child's id, return the
			// component
			if (component.getId() == childId) {
				return component;
			}
		}

		// Nothing was found!
		return null;

	}

	/**
	 * (non-Javadoc)
	 * 
	 * @see Composite#getNumberOfComponents()
	 */
	@Override
	public int getNumberOfComponents() {

		// Return the size
		return this.lWRComponents.size();
	}

	/**
	 * (non-Javadoc)
	 * 
	 * @see Composite#getComponents()
	 */
	@Override
	public ArrayList<Component> getComponents() {
		// Local Declarations
		ArrayList<Component> components = new ArrayList<Component>();
		Component component;
		Iterator<LWRComponent> iter;

		// Setup iterator
		iter = this.lWRComponents.values().iterator();

		// Iterate over the list and get the component
		// Only grab the first in the list
		while (iter.hasNext()) {

			// Iterate the next piece
			component = iter.next();
			// Add the component to the list
			components.add((Component) component);
		}

		// Return the list
		return components;

	}
}